api/desecapi/pdns_change_tracker.py

```diff
@@ -0,0 +1,228 @@
+    - `rr_set_modifications`: set of modified RR sets
+    - `rr_set_deletions`: set of deleted RR sets
+
+    `additions` and `deletions` are guaranteed to be disjoint:
+    - If an item is in the set of additions while being deleted, it is removed from `rr_set_additions`.
+    - If an item is in the set of deletions while being added, it is removed from `rr_set_deletions`.
+    `modifications` and `deletions` are guaranteed to be disjoint.
+    - If an item is in the set of deletions while being modified, an exception is raised.
+    - If an item is in the set of modifications while being deleted, it is removed from `rr_set_modifications`.
+    """
+
+    def __init__(self, optimize=True):
+        self.optimize = optimize
+        self._domain_additions = set()
+        self._domain_deletions = set()
+        self._rr_set_additions = {}
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+
+    def __enter__(self):
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+        post_save.connect(self._on_rr_post_save, sender=RR)
+        post_delete.connect(self._on_rr_post_delete, sender=RR)
+        post_save.connect(self._on_rr_set_post_save, sender=RRset)
+        post_delete.connect(self._on_rr_set_post_delete, sender=RRset)
+        post_save.connect(self._on_domain_post_save, sender=Domain)
+        post_delete.connect(self._on_domain_post_delete, sender=Domain)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_val:
+            return  # TODO proper error treatment
+
+        for domain_name in self._domain_deletions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            assert domain_name not in (self._rr_set_additions.keys()
+                                       | self._rr_set_modifications.keys()
+                                       | self._rr_set_deletions.keys())
+            _pdns_delete(pdns_id)
+
+        for domain_name in self._rr_set_additions.keys() | self._domain_additions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            if domain_name in self._domain_additions:
+                salt = '%016x' % random.randrange(16 ** 16)
+                payload = {'name': normalized_name, 'kind': 'MASTER', 'dnssec': True,
+                           'nsec3param': '1 0 127 %s' % salt, 'nameservers': settings.DEFAULT_NS}
+                _pdns_post('/zones', payload)
+
+            additions = self._rr_set_additions.get(domain_name, set())
+            modifications = self._rr_set_modifications.get(domain_name, set())
+            deletions = self._rr_set_deletions.get(domain_name, set())
+
+            assert additions & deletions == set()
+            assert modifications & deletions == set()
```

Comment:
`assert not (a & b)`

Reply:
Ah, of course empty is False! I'm going to remove the check though.